### PR TITLE
Patient appointments

### PR DESCRIPTION
**Pull request by aliadnanaslam7**
appointments CRUD operations, test cases, validations, authorization

### CODE REVIEW

**Comment by aliadnanaslam7** [comment]
Cycle 0

File: app/views/appointments/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            <td><%= link_to 'Destroy', appointment, method: :delete, data: { confirm: 'Are you sure?' } %></td>
+          <% end %>
+        </tr>
+      <% end %>
+    </tbody>
+  </table>
+</div>
+<br>
+
+<%= link_to 'Request a appointment', new_appointment_path %>
```
Comment (hafsabatool7) [inline app/views/appointments/index.html.erb]: *"Request An Appointment"

File: app/controllers/appointments_controller.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
+class AppointmentsController < ApplicationController
+  before_filter :authenticate_user!
+  before_action :set_appointment, only: %i[show edit update destroy]
```
Comment (hafsabatool7) [inline app/controllers/appointments_controller.rb]: What is the difference before_action and before_filter? Why not used same all over?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,5 +1,26 @@
...
+    Hospital.current_id = nil
+  end
+
+  def authenticate_user!
+    if signed_in?
+      super
+    else
+      redirect_to '/users/sign_in', notice: 'User is not logged in'
+    end
+  end
```
Comment (hafsabatool7) [inline app/controllers/application_controller.rb]: RuboCop error: final newline missing.

File: app/views/patients/dashboard.html.erb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<h1>Hello <%= current_user.name %></h1>
+<h2>Date of birth: <%= current_user.date_of_birth %></h2>
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/patients/dashboard.html.erb]: Nah, it is just for testing purposes. Our dashboard will completely change depending on the theme.

File: app/views/doctors/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        <th>Gender</th>
+        <th colspan="3">Actions</th>
+      </tr>
+    </thead>
+
+    <tbody>
+      <% @doctors.each do |doctor| %>
+        <tr>
+          <td><%= doctor.name %></td>
+          <td>Emergency</td>
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/doctors/index.html.erb]: Since I don't have a doctor module yet, Arslan is working that code. will update this when I get that module.

File: app/views/appointments/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  <%= @appointment.start_time.strftime("%Y-%m-%d") %>
+</p>
+
+<p>
+  <strong>Time slot:</strong>
+  <%= time_slot(@appointment.start_time, @appointment.end_time) %>
+</p>
+
+<p>
+  <strong>Doctor:</strong>
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/appointments/show.html.erb]: done!

File: app/controllers/patients_controller.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class PatientsController < ApplicationController
+  before_filter :authenticate_user!
+  # load_and_authorize_resource
+
+  def dashboard
+    respond_to do |format|
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/patients_controller.rb]: Our team lead said it once we don't leave the empty actions, we should write a respond_to block in the action.

without mentioning the respond_to block, it entertains all the other formats also (.json, .html, .js etc.) Since we are not developing API's or other services, we are only developing a web application and our app should respond to HTML format only. that's why respond_to block is necessary

File: app/controllers/doctors_controller.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+class DoctorsController < ApplicationController
+  before_filter :authenticate_user!
+  before_action :set_doctors, only: :index
+  # load_and_authorize_resource
+
+  def index
+    respond_to do |format|
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/doctors_controller.rb]: done!

File: app/controllers/appointments_controller.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    @appointment.destroy
+    respond_to do |format|
+      format.html { redirect_to appointments_url, notice: 'Appointment was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
+
+  def medical_history
+    @appointments = current_user.appointments.history
+    respond_to do |format|
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: Our team lead said it once we don't leave the empty actions, we should write a respond_to block in the action.

without mentioning the respond_to block, it entertains all the other formats also (.json, .html, .js etc.) Since we are not developing API's or other services, we are only developing a web application and our app should respond to HTML format only. that's why respond_to block is necessary

File: app/views/appointments/medical_history.html.erb [hunk 10]
```diff
@@ -0,0 +1,35 @@
+<p id="notice"><%= notice %></p>
+
+<h1>Appointments</h1>
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/appointments/medical_history.html.erb]: Since prescription and medicine module are not implemented yet. will update when done. For the time being updating the <h1>Appointments</h1> tag.

File: app/controllers/appointments_controller.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        format.json { render :show, status: :ok, location: @appointment }
+      else
+        format.html { render :edit }
+        format.json { render json: @appointment.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def destroy
+    @appointment.destroy
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    end
+  end
+
+  def update
+    respond_to do |format|
+      if @appointment.update(create_appointment_params)
+        format.html { redirect_to @appointment, notice: 'Appointment was successfully updated.' }
+        format.json { render :show, status: :ok, location: @appointment }
+      else
+        format.html { render :edit }
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: this snip ```render json: @appointment.errors``` is sending the error messages(model error messages) to the view through json.

File: app/controllers/appointments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  def create
+    @appointment = Appointment.new(create_appointment_params)
+    respond_to do |format|
+      if @appointment.save
+        format.html { redirect_to @appointment, notice: 'Appointment was successfully created.' }
+        format.json { render :show, status: :created, location: @appointment }
+      else
+        format.html { render :new }
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: this snip render json: @appointment.errors is sending the error messages(model error messages) to the view through json.

File: app/controllers/appointments_controller.rb [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  def create_appointment_params
+    custom_params = appointment_params
+
+    custom_params['doctor_id'] = custom_params['doctor_id'].to_i
+    custom_params['patient_id'] = current_user.id
+    custom_params['hospital_id'] = Hospital.current_id
+
+    custom_params
+  end
+end
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: It shows something like this when final line is missing 
image
Since in this file it is not showing it because there is already an empty line at the end of file.

File: app/controllers/appointments_controller.rb [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+      format.html
+    end
+  end
+
+  def new
+    @appointment = Appointment.new
+  end
+
+  def edit
+    respond_to do |format|
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: Our team lead said it once we don't leave the empty actions, we should write a respond_to block in the action.

without mentioning the respond_to block, it entertains all the other formats also (.json, .html, .js etc.) Since we are not developing API's or other services, we are only developing a web application and our app should respond to HTML format only. that's why respond_to block is necessary

File: app/controllers/appointments_controller.rb [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+  def index
+    current_user.admin? ? @appointments = @hospital.appointments : @appointments = current_user.appointments.requested
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def show
+    respond_to do |format|
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: Our team lead said it once we don't leave the empty actions, we should write a respond_to block in the action.

without mentioning the respond_to block, it entertains all the other formats also (.json, .html, .js etc.) Since we are not developing API's or other services, we are only developing a web application and our app should respond to HTML format only. that's why respond_to block is necessary

File: app/controllers/appointments_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: There is no difference between both, Just checked it, before_filters syntax is old and deprecated.
Updated it to before_action.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (aliadnanaslam7) [reply app/controllers/application_controller.rb]: It shows something like this when final line is missing 
image
Since in this file it is not showing it because there is already an empty line at the end of file.

File: app/controllers/appointments_controller.rb (hunk 2 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: There is no difference between both, Just checked it, before_filters syntax is old and deprecated.
Updated it to before_action.

File: app/views/appointments/index.html.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/appointments/index.html.erb]: done!

File: app/views/devise/sessions/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    <%= f.label :password %><br />
+    <%= f.password_field :password, autocomplete: "off" %>
+  </div>
+
+  <% if devise_mapping.rememberable? -%>
+    <div class="field">
+      <%= f.check_box :remember_me %>
+      <%= f.label :remember_me %>
+    </div>
+  <% end -%>
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/devise/sessions/new.html.erb]: these are the devise generated views and will update with the new UI.

File: app/models/hospital.rb [hunk 18]
```diff
@@ -1,3 +1,4 @@
+# hospital model
 class Hospital < ActiveRecord::Base
   has_many :users, dependent: :destroy #need to ask about dependent: :destory
   has_many :appointments
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/models/hospital.rb]: we are not giving any option to user to delete the hospital, so there is no need to add any dependency.

File: app/models/appointment.rb [hunk 19]
```diff
@@ -1,6 +1,33 @@
...
+  validates :patient_id, presence: true, numericality: { only_integer: true }
+  validates :hospital_id, presence: true, numericality: { only_integer: true }
+  validate  :ensure_start_time_type
+  validate  :ensure_end_time_type
+  validate  :ensure_status
+
+  belongs_to :patient, foreign_key: 'patient_id', class_name: 'User'
+  belongs_to :doctor, foreign_key: 'doctor_id', class_name: 'User'
   belongs_to :hospital
   has_one    :prescription
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: no, if a user deletes an appointment, it does not affect our user.

File: app/models/patient.rb [hunk 20]
```diff
@@ -0,0 +1,4 @@
+# patient model class
+class Patient < User
+  has_many :appointments, class_name: 'Appointment', foreign_key: 'patient_id'
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/models/patient.rb]: done!

File: app/models/doctor.rb [hunk 21]
```diff
@@ -0,0 +1,3 @@
+class Doctor < User
+  has_many :appointments, class_name: 'Appointment', foreign_key: 'doctor_id'
```
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/models/doctor.rb]: done!

File: app/models/appointment.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (hafsabatool7) [reply app/models/appointment.rb]: I was talking about prescription. Since a prescription refers to an appointment, so if the appointment gets deleted, our prescription related to that appointment will be corrupted.

File: app/views/appointments/medical_history.html.erb (hunk 10 above)
[the remark this answers is not in the document]
Comment (hafsabatool7) [reply app/views/appointments/medical_history.html.erb]: Alright.

File: app/views/devise/sessions/new.html.erb (hunk 17 above)
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/views/devise/sessions/new.html.erb]: done!

File: app/models/appointment.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: done!

File: app/models/appointment.rb [hunk 22]
```diff
@@ -1,6 +1,35 @@
 class Appointment < ActiveRecord::Base
-  belongs_to :patient_id, class_name: 'User'
-  belongs_to :doctor_id, class_name: 'User'
+  STATUS = { requested: 'Requested', confirmed: 'Confirmed', treated: 'Treated', cancelled: 'Cancelled' }.freeze
+  sequenceid :hospital, :appointments
+
+  validates :doctor_id, presence: true, numericality: { only_integer: true }
+  validates :patient_id, presence: true, numericality: { only_integer: true }
+  validates :hospital_id, presence: true, numericality: { only_integer: true }
```
Comment (naveed-ahmad) [inline app/models/appointment.rb]: DRY it bro!
```
validates : doctor_id, : patient_id, : hospital_id, presence: true, numericality: { only_integer: true }
```

Comment (aliadnanaslam7) [reply app/models/appointment.rb]: done! 
_thanks_

File: app/controllers/appointments_controller.rb [hunk 23]
```diff
@@ -0,0 +1,96 @@
+class AppointmentsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_appointment, only: %i[show edit update destroy]
+  before_action :set_doctors, only: %i[new create edit update]
+  load_and_authorize_resource
+
+  def index
+    current_user.admin? ? @appointments = @current_hospital.appointments : @appointments = current_user.appointments.requested
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: this should be handles in ability

File: app/controllers/appointments_controller.rb [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  def show
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def new
+    @appointment = current_user.appointments.new
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: same, it should be loaded from ability

File: app/controllers/appointments_controller.rb [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+  def edit
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def create
+    @appointment = current_user.appointments.new(create_appointment_params)
+    respond_to do |format|
+      if @appointment.save
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: current_user.appointments.create(create_appointment_params)

File: app/controllers/appointments_controller.rb [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+        format.html { redirect_to @appointment, notice: 'Appointment was successfully created.' }
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: move message to en.yml

File: app/controllers/appointments_controller.rb [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+        format.html { render :new }
+        format.json { render json: @appointment.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def update
+    respond_to do |format|
+      if @appointment.update(create_appointment_params)
+        format.html { redirect_to @appointment, notice: 'Appointment was successfully updated.' }
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: move all the messaged to en.yml

File: app/controllers/appointments_controller.rb [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+        format.json { head :no_content }
+      else
+        format.html { redirect_to appointments_url, alert: 'There was an error destroying the record.' }
+        format.json { head :no_content }
+      end
+    end
+  end
+
+  def medical_history
+    @appointments = current_user.appointments.history
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: instace variable name should be @history

File: app/controllers/appointments_controller.rb [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  private
+
+  def set_appointment
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: no need for this method
should be handled from cancancan

File: app/controllers/appointments_controller.rb [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  def appointment_params
+    params.require(:appointment).permit(:status, :doctor_id, :start_time, :end_time, :file)
+  end
+
+  def create_appointment_params
+    custom_params = appointment_params
+
+    custom_params['doctor_id'] = custom_params['doctor_id'].to_i
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: why do we need to_i?

File: app/controllers/appointments_controller.rb (hunk 23 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 24 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 25 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: resolved!

File: app/controllers/appointments_controller.rb (hunk 26 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 28 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 29 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 30 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: we don't need it.
resolved!

File: app/controllers/appointments_controller.rb (hunk 27 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/application_controller.rb [hunk 31]
```diff
@@ -38,6 +35,14 @@ def scope_current_hospital
     Hospital.current_id = nil
   end
 
+  def authenticate_user!
+    if signed_in?
+      super
+    else
+      redirect_to '/users/sign_in', notice: 'User is not logged in'
```
Comment (mairamuneer39) [inline app/controllers/application_controller.rb]: move message to en.yml

File: app/controllers/appointments_controller.rb [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  def create
+    respond_to do |format|
+      if @appointment.save
+        format.html { redirect_to @appointment, notice: I18n.t('appointment.created') }
+        format.json { render :show, status: :created, location: @appointment }
+      else
+        format.html { render :new }
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: we are not showing any error message if @appointment is not created

File: app/controllers/appointments_controller.rb [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  def update
+    respond_to do |format|
+      if @appointment.update(appointment_params)
+        format.html { redirect_to @appointment, notice: I18n.t('appointment.updated') }
+        format.json { render :show, status: :ok, location: @appointment }
+      else
+        format.html { render :edit }
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: same
no error message

File: app/controllers/application_controller.rb (hunk 31 above)
Comment (aliadnanaslam7) [reply app/controllers/application_controller.rb]: done!

File: app/controllers/patients_controller.rb [hunk 34]
```diff
@@ -0,0 +1,10 @@
+class PatientsController < ApplicationController
+  before_action :authenticate_user!
+  # load_and_authorize_resource
```
Comment (mairamuneer39) [inline app/controllers/patients_controller.rb]: why is this commented?

File: app/models/appointment.rb [hunk 35]
```diff
@@ -1,6 +1,33 @@
 class Appointment < ActiveRecord::Base
-  belongs_to :patient_id, class_name: 'User'
-  belongs_to :doctor_id, class_name: 'User'
+  STATUS = { requested: 'Requested', confirmed: 'Confirmed', treated: 'Treated', cancelled: 'Cancelled' }.freeze
+  sequenceid :hospital, :appointments
+
+  validates :doctor_id, :patient_id, :hospital_id, presence: true, numericality: { only_integer: true }
+  validate  :ensure_start_time_type
```
Comment (mairamuneer39) [inline app/models/appointment.rb]: better name is
ensure_start_time_type -> validate_start_time

File: app/models/appointment.rb [hunk 36]
```diff
@@ -1,6 +1,33 @@
...
+  validate  :ensure_end_time_type
+  validate  :ensure_status
```
Comment (mairamuneer39) [inline app/models/appointment.rb]: validate_status

File: app/models/appointment.rb [hunk 37]
```diff
@@ -1,6 +1,33 @@
 class Appointment < ActiveRecord::Base
-  belongs_to :patient_id, class_name: 'User'
-  belongs_to :doctor_id, class_name: 'User'
+  STATUS = { requested: 'Requested', confirmed: 'Confirmed', treated: 'Treated', cancelled: 'Cancelled' }.freeze
```
Comment (mairamuneer39) [inline app/models/appointment.rb]: i think ```completed``` is better than treated

File: app/models/appointment.rb [hunk 38]
```diff
@@ -1,6 +1,33 @@
...
+  belongs_to :patient, class_name: 'User', foreign_key: 'patient_id'
+  belongs_to :doctor, class_name: 'User', foreign_key: 'doctor_id'
   belongs_to :hospital
-  has_one    :prescription
+  has_many   :documents, as: :resource
+  has_one    :prescription, dependent: :destroy
+
+  scope :requested, -> { where(status: Appointment::STATUS[:requested]) }
+  scope :history, -> { where(status: Appointment::STATUS[:treated]) }
+  scope :pending, -> { where.not(status: Appointment::STATUS[:treated]) }
```
Comment (mairamuneer39) [inline app/models/appointment.rb]: it will also return cancelled appointments

File: app/models/appointment.rb [hunk 39]
```diff
@@ -1,6 +1,33 @@
...
+
+  private
+
+  def ensure_start_time_type
+    errors[:start_time] << 'must be a valid date' unless start_time.class == ActiveSupport::TimeWithZone
```
Comment (mairamuneer39) [inline app/models/appointment.rb]: en.yml

File: app/models/doctor.rb [hunk 40]
```diff
@@ -1,2 +1,14 @@
 class Doctor < User
+  before_destroy :pending_appointment?
+
+  has_many :appointments, class_name: 'Appointment', foreign_key: 'doctor_id', dependent: :destroy
```
Comment (mairamuneer39) [inline app/models/doctor.rb]: do we need to specify foreign_key?

File: app/models/hospital.rb [hunk 41]
```diff
@@ -1,4 +1,15 @@
+# hospital model
 class Hospital < ActiveRecord::Base
+  has_many :appointments
```
Comment (mairamuneer39) [inline app/models/hospital.rb]: dependent destroy for all associations

File: app/models/medicine.rb [hunk 42]
```diff
@@ -1,3 +1,4 @@
+# medicine model
```
Comment (mairamuneer39) [inline app/models/medicine.rb]: remove this comment

File: app/views/appointments/_form.html.erb [hunk 43]
```diff
@@ -0,0 +1,49 @@
+<%= form_for @appointment, html: { multipart: true } do |f| %>
+<% if @appointment.errors.any? %>
```
Comment (mairamuneer39) [inline app/views/appointments/_form.html.erb]: no need for this

File: app/views/appointments/_form.html.erb [hunk 44]
```diff
@@ -0,0 +1,49 @@
...
+      </div>
+      <div class="form-group col-md-6">
+      <%= f.label :patient_id %>
+      <%= f.select :patient_id, options_from_collection_for_select(@patients, :id, :name, @appointment.patient_id.to_s), { prompt: 'Select a patient' }, { class: "form-control", required: true } %>
+      </div>
+    </div>
+    <div class="form-group">
+      <%= f.label :start_time %>
+      <div class="input-group date" id="appointment_start_time" data-target-input="nearest">
+        <%= f.text_field :start_time, class: "form-control datatimepicker-input", required: true, data: { target: "#appointment_start_time" } %>
```
Comment (mairamuneer39) [inline app/views/appointments/_form.html.erb]: it should be readonly

File: app/views/appointments/_form.html.erb [hunk 45]
```diff
@@ -0,0 +1,49 @@
...
+      <%= f.label :end_time %>
+      <div class="input-group date" id="appointment_end_time" data-target-input="nearest">
+        <%= f.text_field :end_time, class: "form-control datatimepicker-input", required: true, data: { target: "#appointment_end_time" } %>
+        <div class="input-group-append" data-target="#appointment_end_time" data-toggle="datetimepicker">
+          <div class="input-group-text"><i class="fa fa-calendar"></i></div>
+        </div>
+      </div>
+    </div>
+    <div class="form-group">
+      <%= f.label :status %>
```
Comment (mairamuneer39) [inline app/views/appointments/_form.html.erb]: no need to show status in form

File: app/helpers/appointments_helper.rb [hunk 46]
```diff
@@ -0,0 +1,5 @@
+module AppointmentsHelper
+  def time_slot(start_time, end_time)
```
Comment (mairamuneer39) [inline app/helpers/appointments_helper.rb]: format_time is a better name

File: app/views/appointments/medical_history.html.erb [hunk 47]
```diff
@@ -0,0 +1,35 @@
...
+  </thead>
+
+  <tbody>
+    <% @history.each do |appointment| %>
+      <tr>
+        <td><%= appointment.start_time.strftime("%Y-%m-%d") %></td>
+        <td><%= time_slot(appointment.start_time, appointment.end_time) %></td>
+        <td><%= appointment.doctor.name %></td>
+        <td><%= appointment.status.capitalize %></td>
+        <td><%= link_to 'Show', appointment %></td>
```
Comment (mairamuneer39) [inline app/views/appointments/medical_history.html.erb]: check for ability

File: app/views/devise/unlocks/new.html.erb [hunk 48]
```diff
@@ -13,4 +13,5 @@
   </div>
 <% end %>
 
-<%= render "User/shared/links" %>
+<%= render "devise/shared/links" %>
+<%#= render "User/shared/links" %>
```
Comment (mairamuneer39) [inline app/views/devise/unlocks/new.html.erb]: remove if it is not required

File: app/views/doctors/index.html.erb [hunk 49]
```diff
@@ -0,0 +1,28 @@
...
+        <th>Fee</th>
+        <th>Gender</th>
+        <th colspan="3">Actions</th>
+      </tr>
+    </thead>
+
+    <tbody>
+      <% @doctors.each do |doctor| %>
+        <tr>
+          <td><%= doctor.name %></td>
```
Comment (mairamuneer39) [inline app/views/doctors/index.html.erb]: it should be link to detail page

File: app/views/documents/_form.html.erb [hunk 50]
```diff
@@ -0,0 +1,25 @@
+<%= form_for @document, html: { multipart: true } do |f| %>
+<% if @document.errors.any? %>
```
Comment (mairamuneer39) [inline app/views/documents/_form.html.erb]: no need for this

File: app/views/documents/index.html.erb [hunk 51]
```diff
@@ -0,0 +1,38 @@
...
+    </thead>
+
+    <tbody>
+      <% @documents.each do |document| %>
+        <tr>
+          <td><%= document.title %></td>
+          <td><%= document.created_at.strftime('%Y-%m-%d %I:%M') %></td>
+          <td><%= number_to_human_size(document.attachment_file_size) %></td>
+          <td><%= document.attachment_content_type %></td>
+          <td><%= link_to 'Show', document %></td>
```
Comment (mairamuneer39) [inline app/views/documents/index.html.erb]: check for ability

File: db/migrate/20190715185606_devise_create_users.rb [hunk 52]
```diff
@@ -1,43 +1,43 @@
...
-
-
-      t.timestamps
-    end
-
-    add_index :users, :email,                unique: true
-    add_index :users, :reset_password_token, unique: true
-    add_index :users, :confirmation_token,   unique: true
-    add_index :users, :unlock_token,         unique: true
+    # create_table(:users) do |t|
```
Comment (mairamuneer39) [inline db/migrate/20190715185606_devise_create_users.rb]: revert this change

File: db/migrate/20190715190204_create_hospitals.rb [hunk 53]
```diff
@@ -1,13 +1,13 @@
 class CreateHospitals < ActiveRecord::Migration
   def change
-    create_table :hospitals do |t|
-      t.string :name
-      t.string :subdomain
+    # create_table :hospitals do |t|
```
Comment (mairamuneer39) [inline db/migrate/20190715190204_create_hospitals.rb]: same

File: db/migrate/20190725143606_create_documents.rb [hunk 54]
```diff
@@ -0,0 +1,13 @@
+class CreateDocuments < ActiveRecord::Migration
+  def change
+    create_table :documents do |t|
+      t.integer    :hospital_id
```
Comment (mairamuneer39) [inline db/migrate/20190725143606_create_documents.rb]: index

File: db/migrate/20190725143606_create_documents.rb [hunk 55]
```diff
@@ -0,0 +1,13 @@
...
+      t.integer    :user_id
```
Comment (mairamuneer39) [inline db/migrate/20190725143606_create_documents.rb]: index

**Comment by mairamuneer39** [comment]
cycle 1

File: db/migrate/20190715190204_create_hospitals.rb (hunk 53 above)
Comment (aliadnanaslam7) [reply db/migrate/20190715190204_create_hospitals.rb]: done!

File: db/migrate/20190715185606_devise_create_users.rb (hunk 52 above)
Comment (aliadnanaslam7) [reply db/migrate/20190715185606_devise_create_users.rb]: done!

File: app/views/devise/unlocks/new.html.erb (hunk 48 above)
Comment (aliadnanaslam7) [reply app/views/devise/unlocks/new.html.erb]: removed!

File: app/helpers/appointments_helper.rb (hunk 46 above)
Comment (aliadnanaslam7) [reply app/helpers/appointments_helper.rb]: done!

File: app/views/appointments/_form.html.erb (hunk 44 above)
Comment (aliadnanaslam7) [reply app/views/appointments/_form.html.erb]: it disables the whole field and can't take date input.

File: app/models/medicine.rb (hunk 42 above)
Comment (aliadnanaslam7) [reply app/models/medicine.rb]: removed!

File: app/models/appointment.rb (hunk 39 above)
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: done!

File: app/models/appointment.rb (hunk 38 above)
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: resolved!

File: app/models/appointment.rb (hunk 37 above)
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: done!

File: app/models/appointment.rb (hunk 36 above)
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: done!

File: app/models/appointment.rb (hunk 35 above)
Comment (aliadnanaslam7) [reply app/models/appointment.rb]: done!

File: app/controllers/patients_controller.rb (hunk 34 above)
Comment (aliadnanaslam7) [reply app/controllers/patients_controller.rb]: It's needed for the time being.

File: app/models/doctor.rb (hunk 40 above)
Comment (aliadnanaslam7) [reply app/models/doctor.rb]: no, it's working without it.
removed!

File: app/models/hospital.rb (hunk 41 above)
Comment (aliadnanaslam7) [reply app/models/hospital.rb]: done!

File: app/views/documents/index.html.erb (hunk 51 above)
Comment (aliadnanaslam7) [reply app/views/documents/index.html.erb]: done!

File: app/models/hospital.rb (hunk 41 above)
Comment (aliadnanaslam7) [reply app/models/hospital.rb]: done!

File: db/migrate/20190725143606_create_documents.rb (hunk 55 above)
Comment (aliadnanaslam7) [reply db/migrate/20190725143606_create_documents.rb]: added

File: app/views/appointments/medical_history.html.erb (hunk 47 above)
Comment (aliadnanaslam7) [reply app/views/appointments/medical_history.html.erb]: changed!
